server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -490,64 +493,45 @@ public void clear() {
 
         @Override
         public int size() {
-            return size;
+            return values.size();
         }
 
         @Override
-        public Iterator iterator() {
-            return new Iterator<>() {
-                private final StreamInput in;
-                private int pos = 0;
-
-                {
-                    try {
-                        in = valueRef.streamInput();
-                        consumerHeadersAndGetListSize(in);
-                    } catch (IOException e) {
-                        throw new UncheckedIOException("failed to deserialize TermsQueryBuilder", e);
-                    }
-                }
-
-                @Override
-                public boolean hasNext() {
-                    return pos < size;
-                }
-
-                @Override
-                public Object next() {
-                    try {
-                        pos++;
-                        return in.readGenericValue();
-                    } catch (IOException e) {
-                        throw new UncheckedIOException("failed to deserialize TermsQueryBuilder", e);
-                    }
-                }
-            };
+        public Iterator<?> iterator() {
+            var iter = values.iterator();
+            return convert ? Iterators.map(iter, AbstractQueryBuilder::maybeConvertToBytesRef) : iter;
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
-            out.writeBytesReference(valueRef);
+            out.writeBytesReference(asBytes());
+        }
+
+        private BytesReference asBytes() {
```

Comment:
this caching makes me nervous. Is there never a situation where the same query builder instance is shared across different threads? Sounds like an edge case at least, but couldn't it theoretically happen?

Reply:
> Is there never a situation where the same query builder instance is shared across different threads? Sounds like an edge case at least, but couldn't it theoretically happen?

I think practically this is not something that happens right now. But even if it were to happen, we always build the bytes before we assign them. So worst case, we would serialise more often than we need to which is IMO much preferable to introducing a volatile that adds yet more hard to understand memory effects?